### PR TITLE
New more convenient ViewerApp API

### LINKED ISSUE #146: New Viewer API
## 🚀 Feature

As described in #120 we will move the new API towards the following

```python
from napari import Viewer
viewer = Viewer(astro)
viewer.add_image(camera)
```

and

```python
from napari import Viewer
viewer = Viewer()
viewer.add_image(astro)
viewer.add_image(camera)
```
where the input arguments to Viewer get treated identically to the input arguments to add_image as images are our main layer type and we want away to show an image in 2-lines.

Additionally now window creation will be handled inside Viewer and the window will be accessible through viewer.window and the layers will be accessible through viewer.layers. Under the hood though we will using a pass-through from a Viewer object to a ViewerWidget object (that does not know about Window) such that the following code path would also be acceptable for advanced developers that want to embed Napari in their own applications.

```python
from napari import ViewerWidget
viewer = ViewerWidget()
viewer.add_image(astro)
viewer.add_image(camera)
custom_application.add(viewer)
```

### PR DESCRIPTION
**Pull request by jni**
# Description
This creates a new class that builds both a Viewer and a Window in one fell swoop. This is the fancy object that we will use as the preferred entry point to Napari, with the current Viewer to be used as a Widget that could be e.g. embedded in other applications.

To be clear, the new name is provisional while @royerloic messes with the Viewer. After both his and my PR are merged, the plan is to refactor:

Viewer -> ViewerWidget
ViewerApp -> Viewer

with the latter name up for debate.

All kinds of comments and criticisms of the new API are welcome.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# References
Fixes #146 

# How has this been tested?

- [x] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: napari/viewer.py
Comment (sofroniewn) [inline napari/viewer.py]: I like being able to pass either one image or the named images, that's very nice. Do we also want the ability to set the keyword arguments of `add_image` too, such as the meta data? As a separate PR I think we want to make things like `clims`, `colormap` etc. settable in the constructor of `add_image` and I think we want to be able to just propagate any of those things through this part of the API too

File: examples/nD_image.py
Comment (sofroniewn) [inline examples/nD_image.py]: I get a whole bunch of errors / warnings with this example now when I move the mouse over the image or adjust the clims

```python
  File "/Users/nicholassofroniew/Github/napari/napari/layers/_image_layer/model.py", line 387, in get_value
    msg = msg + f'{value:0.3}'
ValueError: Precision not allowed in integer format specifier
ERROR: Invoking <bound method QtViewer.on_mouse_move of <napari.components._viewer.view.main.QtViewer object at 0x111b10ca8>> for MouseEvent
WARNING: Traceback (most recent call last):
  File "/Users/nicholassofroniew/Github/napari/napari/layers/_image_layer/view/controls.py", line 33, in clim_slider_changed
    cmin = valmin+slidermin*(valmax-valmin)
TypeError: numpy boolean subtract, the `-` operator, is deprecated, use the bitwise_xor, the `^` operator, or the logical_xor function instead.
```

Not clear to me that the blobs are more visually appealing / clearer to look at then the random noise. Same is true for the `nD_markers.py` example.

File: examples/set_colormaps.py
Comment (sofroniewn) [inline examples/set_colormaps.py:19]: In line 8 of this file, should the import now just be `import napari` not `import napari.util`? It doesn't seem to make a difference which confuses me.

Comment (sofroniewn) [reply examples/set_colormaps.py:19]: I have changed it to `import napari` as it looks cleaner to me, hope that's ok

**Comment by sofroniewn** [comment]
I have pushed a commit to fix the notebook (which still had `napari_gui` in there!!)

**Review by sofroniewn — APPROVED** [review]
This looks good to me (though see my comments on pass through of the `add_image` keyword args and the new nD examples).

Moving the names over to `ViewerWidget` and `Viewer` after the dims refactor lands sounds reasonable to me.

File: examples/nD_image.py
Comment (jni) [reply examples/nD_image.py]: > Not clear to me that the blobs are more visually appealing / clearer to look at

Huh? it's night and day for me, you get some visual continuity between the 3D planes!

Anyway, it seems we can't display boolean images, easy enough fix.

**Comment by jni** [comment]
Why is Windows failing? @kne42, any ideas?

File: napari/viewer.py
Comment (jni) [reply napari/viewer.py]: I did think about that, I'm happy to do that, but here are the caveats:

- any named parameters will be forbidden as image names
- the same parameters will apply to all images. I actually started down the rabbit hole of passing a list or dictionary of configs, but that got real ugly real fast, and I decided against it.

Comment (jni) [reply napari/viewer.py]: Side note, clims needs a name change

File: examples/nD_image.py
Comment (sofroniewn) [reply examples/nD_image.py]: Fair enough, thanks for the fix. We can work on the data for examples in future PRs!

File: napari/viewer.py
Comment (sofroniewn) [reply napari/viewer.py]: Hmm you make good points. I agree that the dictionary approach is not too appealing, but I don't think I mind the extra keyword args applying to all image layers. I view this as convenience functionality and as such might have to be used with some caution.

It would be a shame to lose the ability to have named images completely as you do in the current examples, but it also doesn't really bother me to have a couple "protected" keywords that cannot be used for names, such as `colormap` etc.

Curious what @kne42 thinks

Comment (kne42) [reply napari/viewer.py]: I think as long as you "reserve" the keyword arguments by explicitly including them in the signature definition, it should be clear enough that it won't get too confusing.

File: napari/viewer.py
Comment (kne42) [inline napari/viewer.py]: docstringssss

File: napari/viewer.py
Comment (kne42) [inline napari/viewer.py]: docstringsssss

File: napari/viewer.py
Comment (kne42) [inline napari/viewer.py:5]:
```suggestion
    """napari ndarray viewer.
```
🤓

**Review by kne42 — COMMENTED** [review]
Kind of on the fence on the implementation with regards to effectively wrapping some aspects of the `ViewerWidget` with the `ViewerApp`. I like that this cleanly separates convenience functionality but I also am not a huge fan of wrapping all of the internals. Say for example someone wants to share a particular region of an image. To set the view they would have to use the workflow `viewer.viewer_widget.camera.set_range(region)` which I find kinda messy, redundant, and otherwise overly complex.

File: napari/viewer.py
Comment (jni) [reply napari/viewer.py:5]: The jury is still out on this one, and I think we should leave this discussion until Mar 29. ;)

**Comment by jni** [comment]
@kne42 

> but I also am not a huge fan of wrapping all of the internals. Say for example someone wants to share a particular region of an image. To set the view they would have to use the workflow `viewer.viewer_widget.camera.set_range(region)` which I find kinda messy, redundant, and otherwise overly complex.

Oh yeah that's terrible syntax. The idea would be to wrap the smallest amount possible. My suggestion would be to pass through `layers` so as to get a `LayerList`, and then you can do something like `viewer.layers['camera'].set_range(region)`. I think that's fine. Ultimately, we can't both not wrap viewer and not have a Window object, and I think not having to deal with a Window object takes precedence.

I've added support for the existing parameters to `add_image`. Currently `colormap=` and `interpolation=` are not supported, but I think that can wait for a future PR.

File: napari/viewer.py
Comment (jni) [reply napari/viewer.py:5]: Also, I believe the entire repo is a hodgepogde of this stuff, so again, fixes can come all together later.

**Comment by sofroniewn** [comment]
Ok this all looks reasonable to me and will merge if @kne42 approves. We can revisit the pass through (and the capital letters) in the future

**Comment by kne42** [comment]
What do you think about just inheriting `Viewer` instead? This way we don't need to wrap anything at all.

Also, I think that neither of these cases address what we should do in the case that we have another tab of window (such as a tabbed window). Should the convenience API be made solely for this type of window or should it be placed on the viewer object for common access instead?

**Comment by jni** [comment]
> What do you think about just inheriting Viewer instead?

I'm not sure! Could work!

> Should the convenience API be made solely for this type of window or should it be placed on the viewer object for common access instead?

Also not sure, but I think probably this is a window thing (one window -> multiple viewerwidgets). This might argue against your above suggestion...?

**Comment by jni** [comment]
Ok this is ready for a (one hopes) final pass. ;)

**Comment by sofroniewn** [comment]
This all looks good to me!

**Review by kne42 — APPROVED** [review]
LGTM! :)

**Pull request merged**
